Prevent overflow on StatsCollector.java

GabrielBrascher opened the pull request:
## Description
<!--- Describe your changes in detail -->
Did a quick check and found an operation that can result in an overflow. Additionally, I took the liberty to do a couple of code enhancements/clean ups.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

GabrielBrascher commented on server/src/main/java/com/cloud/server/StatsCollector.java:
```diff
@@ -454,8 +456,9 @@ protected void init(Map<String, String> configs) {
             s_logger.warn("Usage stats job aggregation range is to small, using the minimum value of " + UsageUtils.USAGE_AGGREGATION_RANGE_MIN);
             _usageAggregationRange = UsageUtils.USAGE_AGGREGATION_RANGE_MIN;
         }
-        _diskStatsUpdateExecutor.scheduleAtFixedRate(new VmDiskStatsUpdaterTask(), (endDate - System.currentTimeMillis()), (_usageAggregationRange * 60 * 1000),
```
Potential overflow in 2 Values before it is converted to long by use in an invocation context.

GabrielBrascher commented on server/src/main/java/com/cloud/server/StatsCollector.java:
```diff
@@ -1209,7 +1212,7 @@ private String getAutoscaleAction(HashMap<Long, Double> avgCounter, long groupId
             for (AutoScaleVmGroupPolicyMapVO asVmgPmap : listMap) {
                 AutoScalePolicyVO policyVO = _asPolicyDao.findById(asVmgPmap.getPolicyId());
                 if (policyVO != null) {
-                    Integer quitetime = policyVO.getQuietTime();
```
The variable 'quitetime' is only assigned values of primitive type and is never 'null', but it is declared with the boxed type 'Integer'.

GabrielBrascher commented on server/src/main/java/com/cloud/server/StatsCollector.java:
```diff
@@ -1533,7 +1536,7 @@ protected void writeBatches(InfluxDB influxDbConnection, String dbName, List<Poi
      * The considered disk stats are the following: bytes read, bytes write, IO read,  and IO write.
      */
     protected boolean isCurrentVmDiskStatsDifferentFromPrevious(VmDiskStatisticsVO previousVmDiskStats, VmDiskStatisticsVO currentVmDiskStats) {
-        if (previousVmDiskStats != null) {
```
currentVmDiskStats might be null as well.

ustcweizhou replied:
@GabrielBrascher it looks currentVmDiskStats is not null


https://github.com/apache/cloudstack/blob/9a1486fdebc3d4cfd17d581ec1b66fbd9c3a4bcd/server/src/main/java/com/cloud/server/StatsCollector.java#L735-L746

GabrielBrascher replied:
Just updated the code addressing your comment, @ustcweizhou. Thanks for the heads up :+1:

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-1000

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1191)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 28797 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3932-t1191-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_ssvm.py
Smoke tests completed. 81 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

GabrielBrascher commented:
Thanks for all the reviews!

Just to mention. I did a few manual tests to ensure that no overflow will happen. Setting ONE_MINUTE_IN_MILLISCONDS to Long (or long) will indeed solve overflow issues.

DaanHoogland commented:
@GabrielBrascher as this is a pure fix, will it apply to 4.13?

GabrielBrascher commented:
@borisstoyanov the overflow is not something that will happen easily; however, the code is there and it still something that might happen one day depending on the value stored at `_usageAggregationRange`.

The operation that I extracted can result in an overflow as it is an **integer** operation stored in an **integer** _memory_ (and then after storing it as an integer it is cast into a long). I extracted it into a `long = int * Long`, which prevents a overflow as it is not an integer operation anymore but converted into a long operation.

GabrielBrascher commented:
@DaanHoogland I see no problem in applying this at 4.13.1.

DaanHoogland commented:
ok, can you rebase @GabrielBrascher ?

GabrielBrascher commented:
@DaanHoogland  rebased it to 4.13 :slightly_smiling_face:

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-1037

borisstoyanov commented:
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1225)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 33837 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3932-t1225-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_internal_lb.py
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Smoke tests completed. 76 look OK, 1 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_vpc_privategw_static_routes | `Failure` | 219.74 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 210.20 | test_privategw_acl.py
test_04_rvpc_privategw_static_routes | `Failure` | 942.54 | test_privategw_acl.py

andrijapanicsb commented:
2 x LGTMs, all tests passing (those privategtw tests are "OK", fixed in newer ACS code - this PR is 150+ commits behind master)

Merging.
